text/0000-overload-logic-operators.md

```diff
@@ -0,0 +1,211 @@
+- Feature Name: overload-logic-operators
+- Start Date: 2019-07-04
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This feature would allow for the two short circuit operators `||` and `&&` to be overloadable by
+users-of-rust and for the standard library (probably in core/alloc) to implement such an overload for
+the `Option<T>` and `Result<T, E>` types for `||` and for the `Option<T>` type `&&` but not for
+`Result<T, E>`.
+
+# Motivation
+[motivation]: #motivation
+
+This idea was original floated as a way to clear up the differences between `.or(...)`, `.or_with(|| ...)`, `.and(...)`, `.and_with(|| ...)`, `.unwrap_or(...)`, and `.unwrap_or_with(|| ...)`. Not only was the requirement to remember that there were additional methods that are supposed to be used when you don't want to compute the value before the check (short circuiting). There was also a concern about the overhead of the additional closure.
+
+This proposal is mostly about reducing the mental strain when chaining `Option<T>`'s and `Result<T, E>`'s. But has a very nice side effect of allowing users-of-rust the ability to overload these operators.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This proposal starts with an enum definition and trait definitions for each of the operators:
+
+```rust
+enum ShortCircuit<S, L> {
+    Short(S),
+    Long(L),
+}
+
+trait LogicalOr<Rhs = Self>: Sized {
+    type Output;
+
+    /// Decide whether the *logical or* should short-circuit
+    /// or not based on its left-hand side argument. If so,
+    /// return its final result, otherwise return the value
+    /// that will get passed to `logical_or()` (normally this
+    /// means returning self back, but you can change the value).
+    fn short_circuit_or(self) -> ShortCircuit<Self::Output, Self>;
+
+    /// Complete the *logical or* in case it did not short-circuit.
+    /// Normally this would just return `rhs`.
+    fn logical_or(self, rhs: Rhs) -> Self::Output;
```

Comment:
These definitions seem to discard information, so seem like they'd be less than optimal when writing an impl.

For example, if `short_circuit_or` returns `Short` for an `Ok`, then `logical_or` still gets passed the whole `Result`, and would need to do `unimplemented!()` or something in the `Ok` arm instead of only being passed the `Err` part.

If it were, instead, `-> ShortCircuit<Self::Short, Self::Long>`, then it could be `logical_or(Self::Long, Rhs)`.  But of course then `short_circuit` becomes exactly the same as `Try::into_result`...

Reply:
If we have a distinct long type (intermediate) then the `short_circuit` for `Result` would still just be `rhs`.